### ISSUE TITLE
Strange incompatibility between biblatex and refcheck

### ISSUE THREAD

**Issue by dliessi** [plk/biblatex#1070]
Consider this example:
```
\documentclass{article}

\usepackage{hyperref}

\usepackage{biblatex}
\DeclareBibliographyCategory{b}
\addtocategory{b}{a}

\usepackage{refcheck}
\makeatletter
\AtBeginDocument{%
  \@ifpackageloaded{refcheck}{%
    \@ifundefined{hyperref}{}{%
      \let\T@ref@orig\T@ref
      \def\T@ref#1{\T@ref@orig{#1}\wrtusdrf{#1}}%
      \let\@refstar@orig\@refstar
      \def\@refstar#1{\@refstar@orig{#1}\wrtusdrf{#1}}%
      \DeclareRobustCommand\ref{\@ifstar\@refstar\T@ref}%
    }%
  }{}%
}
\makeatother

\begin{document}
\begin{equation}\label{K}
K
\end{equation}
\ref*{K}
\end{document}
```

The `\ref*{K}` command is not recognized correctly and results in `??K`.
Removing either the biblatex-related block (or only the category-related lines) or the refcheck-related block in the preamble results in the desired behaviour, i.e. `\ref*{K}` results in `1`.
This happens with biblatex 3.15a (with the fix from 2e5ec8bd1d267203b76a2a6922507ebd498af9e2 applied manually), while it did not happen with biblatex 3.14.
The lines between the `\makeat...` pair solve an incompatibiliy between refcheck and hyperref (workaround from https://tex.stackexchange.com/questions/273970/conflict-refcheck-subcaption-packages-for-label-with-underscores).

I'm using a fully updated TeX Live 2020 on Fedora 33.

**Comment by dliessi** [comment]
> Consider this example:
> 
> ```
> \documentclass{article}
> 
> \usepackage{hyperref}
> 
> \usepackage{biblatex}
> \DeclareBibliographyCategory{b}
> \addtocategory{b}{a}
> 
> \usepackage{refcheck}
> \makeatletter
> \AtBeginDocument{%
>   \@ifpackageloaded{refcheck}{%
>     \@ifundefined{hyperref}{}{%
>       \let\T@ref@orig\T@ref
>       \def\T@ref#1{\T@ref@orig{#1}\wrtusdrf{#1}}%
>       \let\@refstar@orig\@refstar
>       \def\@refstar#1{\@refstar@orig{#1}\wrtusdrf{#1}}%
>       \DeclareRobustCommand\ref{\@ifstar\@refstar\T@ref}%
>     }%
>   }{}%
> }
> \makeatother
> 
> \begin{document}
> \begin{equation}\label{K}
> K
> \end{equation}
> \ref*{K}
> \end{document}
> ```
> 
> The `\ref*{K}` command is not recognized correctly and results in `??K`.
> The lines between the `\makeat...` pair solve an incompatibiliy between refcheck and hyperref (workaround from https://tex.stackexchange.com/questions/273970/conflict-refcheck-subcaption-packages-for-label-with-underscores).

I forgot to mention that also this incompatibility is related to `\ref*`.
As I wrote above, with biblatex 3.14 that workaround is enough to make `\ref*` work with refcheck.

**Comment by hvoss49** [comment]
Works for me if I change the order of the packages:

```
\usepackage{hyperref}
\usepackage{refcheck}
\makeatletter
\AtBeginDocument{%
  \@ifpackageloaded{refcheck}{%
    \@ifundefined{hyperref}{}{%
      \let\T@ref@orig\T@ref
      \def\T@ref#1{\T@ref@orig{#1}\wrtusdrf{#1}}%
      \let\@refstar@orig\@refstar
      \def\@refstar#1{\@refstar@orig{#1}\wrtusdrf{#1}}%
      \DeclareRobustCommand\ref{\@ifstar\@refstar\T@ref}%
    }%
  }{}%
}
\makeatother

\usepackage{biblatex}
\DeclareBibliographyCategory{b}
\addtocategory{b}{a}

```

**Comment by dliessi** [comment]
I thought I had tried it, but apparently I had not: indeed changing the order solves my issue, also in the full document.
@hvoss49 thanks for your help!

Is it worth adding a note in the documentation?
I leave the issue open because of this, but I consider my problem solved.

**Comment by moewew** [comment]
The MWE seems to work fine again with 3.16-dev (the current `dev` branch), so I'd say we're probably good here.

**Comment by moewew** [comment]
Note that isn't fully `biblatex`'s fault. Here is a similar MWE that reproduces the error without `biblatex`.
```latex
\documentclass{article}

\AtBeginDocument{\relax}

\usepackage{hyperref}
\usepackage{refcheck}


\makeatletter
\AtBeginDocument{%
  \@ifpackageloaded{refcheck}{%
    \@ifundefined{hyperref}{}{%
      \let\T@ref@orig\T@ref
      \def\T@ref#1{\T@ref@orig{#1}\wrtusdrf{#1}}%
      \let\@refstar@orig\@refstar
      \def\@refstar#1{\@refstar@orig{#1}\wrtusdrf{#1}}%
      \DeclareRobustCommand\ref{\@ifstar\@refstar\T@ref}%
    }%
  }{}%
}
\makeatother

\LogHook{begindocument}

\begin{document}
\begin{equation}\label{K}
K
\end{equation}
\ref*{K}
\end{document}
```

The 'problem' is that the new hook management system collects hooks with the same label and you can no longer guarantee sequential execution of `\AtBeginDocument` calls.
In particular the execution order in this case is
```
> Execution order:
>     top-level, hycolor, hyperref, refcheck.
```
which means that the `hyperref`/`\ref*` code is executed before the hook code from `refcheck` and `hyperref`, which means that it doesn't do what it is supposed to.

With `biblatex` the problem arises because it issues an unlabelled call to `\AtBeginDocument` that goes to `top-level` and thus works like the `\AtBeginDocument{\relax}` in the MWE. The clash with `biblatex` specifically should be resolved in https://github.com/plk/biblatex/issues/1066 (which is why your MWE works with `biblatex` 3.16-dev), but there is no guarantee that you won't run into a similar issue with a different package or different code.

I'm not very familiar with the new LaTeX hook system, so take this with a shovel of salt, but you could avoid these issues by explicitly setting up a rule for your code. (Indeed in the MWE it is enough to use a labelled hook and the rules aren't needed, but you shouldn't rely on that: Only the rules will guarantee the correct order of execution.)

```latex
\documentclass{article}

\AtBeginDocument{\relax}

\usepackage{hyperref}
\usepackage{refcheck}


\makeatletter
\AtBeginDocument[makerefcheckacceptrefstar]{%
  \@ifpackageloaded{refcheck}{%
    \@ifundefined{hyperref}{}{%
      \let\T@ref@orig\T@ref
      \def\T@ref#1{\T@ref@orig{#1}\wrtusdrf{#1}}%
      \let\@refstar@orig\@refstar
      \def\@refstar#1{\@refstar@orig{#1}\wrtusdrf{#1}}%
      \DeclareRobustCommand\ref{\@ifstar\@refstar\T@ref}%
    }%
  }{}%
}
\makeatother

\DeclareHookRule{begindocument}{makerefcheckacceptrefstar}{after}{hyperref}
\DeclareHookRule{begindocument}{makerefcheckacceptrefstar}{after}{refcheck}

\LogHook{begindocument}

\begin{document}
\begin{equation}\label{K}
K
\end{equation}
\ref*{K}
\end{document}
```

**Comment by moewew** [comment]
As explained above, the `biblatex` component of this issue should be resolved in v3.16 (released 31 Dec 2020 and available now in both TeX live and MIkTeX). At the same time, the LaTeX team worked on making `\AtBeginDocument` slightly more reliable (https://github.com/latex3/latex2e/issues/441), so the code example that reproduced the error without `biblatex` works fine if you have LaTeX 2020-10-01 PL3 (also released 31 Dec 2020).

In any case, the problem should disappear if you update your TeX system.
